lib/krb5/pkinit.c

```diff
@@ -829,33 +829,39 @@ pk_verify_sign(krb5_context context,
 	       struct krb5_pk_cert **signer)
 {
     hx509_certs signer_certs;
-    int ret, flags = 0;
+    int ret;
+    unsigned flags = 0, verify_flags = 0;
+
+    *signer = NULL;
 
-    /* BTMM is broken in Leo and SnowLeo */
     if (id->flags & PKINIT_BTMM) {
 	flags |= HX509_CMS_VS_ALLOW_DATA_OID_MISMATCH;
 	flags |= HX509_CMS_VS_NO_KU_CHECK;
 	flags |= HX509_CMS_VS_NO_VALIDATE;
     }
+    if (id->flags & PKINIT_NO_KDC_ANCHOR)
+	flags |= HX509_CMS_VS_NO_VALIDATE;
 
-    *signer = NULL;
-
-    ret = hx509_cms_verify_signed(context->hx509ctx,
-				  id->verify_ctx,
-				  flags,
-				  data,
-				  length,
-				  NULL,
-				  id->certpool,
-				  contentType,
-				  content,
-				  &signer_certs);
+    ret = hx509_cms_verify_signed_ext(context->hx509ctx,
+				      id->verify_ctx,
+				      flags,
+				      data,
+				      length,
+				      NULL,
+				      id->certpool,
+				      contentType,
+				      content,
+				      &signer_certs,
+				      &verify_flags);
     if (ret) {
 	pk_copy_error(context, context->hx509ctx, ret,
 		      "CMS verify signed failed");
 	return ret;
     }
 
+    if ((verify_flags & HX509_CMS_VSE_VALIDATED) == 0)
+	goto out;
```

Comment:
But `ret == 0` here...  Don't we need to set it to something?  Isn't this an error condition?

Reply:
I think this needs fixing, otherwise this looks good!

Reply:
It's valid to return 0 but no signing certificate if `HX509_CMS_VSE_VALIDATED` was set. See the caller assert below:

```c
    heim_assert(host || (ctx->id->flags & PKINIT_NO_KDC_ANCHOR),
                "KDC signature must be verified unless PKINIT_NO_KDC_ANCHOR set");
```